blocks/datasets/schemes.py

```diff
@@ -102,27 +135,48 @@ class SequentialScheme(BatchScheme):
     The batch size isn't enforced, so the last batch could be smaller.
 
     """
-    def __init__(self, num_examples, batch_size):
-        self.num_examples = num_examples
-        self.batch_size = batch_size
+    def get_request_iterator(self):
+        return BatchIterator(self.num_examples, self.batch_size)
+
+
+class ShuffledScheme(BatchScheme):
+    """Shuffled batches iterator.
+
+    Iterate over all the examples in a dataset of fixed size in shuffled
+    batches.
+
+    Notes
+    -----
+    The batch size isn't enforced, so the last batch could be smaller.
+
+    """
+    def __init__(self, *args, **kwargs):
+        self.rng = kwargs.pop('rng', None)
+        super(ShuffledScheme, self).__init__(*args, **kwargs)
 
     def get_request_iterator(self):
-        return SequentialIterator(self.num_examples, self.batch_size)
+        return BatchIterator(self.num_examples, self.batch_size,
+                             shuffled=True, rng=self.rng)
 
 
-class SequentialIterator(six.Iterator):
-    def __init__(self, num_examples, batch_size):
-        self.num_examples = num_examples
+class BatchIterator(six.Iterator):
+    def __init__(self, num_examples, batch_size, shuffled=False, rng=None):
+        indices = numpy.arange(num_examples)
```

Comment:
This will hurt if this iterator is ever used with a large out-of-memory dataset. It is rather unlikely that such datasets will support indexing at all, but when I said that at NIPS you had a counter-example in mind.